### DISCUSSION TITLE
How to mount Ventoy extfs partition r/w in FreeBSD?

### DISCUSSION THREAD

**Discussion by probonopd** [ventoy/Ventoy#1277] [Q&A]
By default the Ventoy partition gets mounted read-only in FreeBSD:

```
% sudo mount.exfat -o uid=0 -o gid=0 -o umask=002 -o noatime /dev/ada1p1 /media/Ventoy
FUSE exfat 1.3.0
WARN: '/dev/ada1p1' is write-protected, mounting read-only.
WARN: volume was not unmounted cleanly.
```

`mount.exfat` says `volume was not unmounted cleanly`, presumably because the operating system is running from it. Because of this, `'/dev/ada1p1` cannot be mounted read-write.

How can we get the exfat volume that the ISO is booted from to mount read-write?

(I tested this with the ISO from https://github.com/helloSystem/ISO/releases/tag/experimental-13.0 that comes with the Ventoy kernel module preinstalled.)

**Comment by ventoy** [comment]
This is a limit of FreeBSD geom framework.
If you create a  device use `gconcat` then the new device will make the target device busy and can not be mounted in rw mode.
It's the same case with `gventoy` driver.

**Reply by probonopd** [reply]
Thanks, can you imagine that the FreeBSD developers could make some change so that it could work?

**Reply by ventoy** [reply]
I don't think FreeBSD developers will change. 
But I have an idea to make some change in `geom_ventoy` driver may be it can work but you need to do a test.

https://github.com/ventoy/Ventoy/blob/master/Unix/ventoy_unix_src/FreeBSD/geom_ventoy_src/13.x/sys/geom/ventoy/g_ventoy.c
Download the source and modify `g_ventoy_access` function as follows:
```
static int
g_ventoy_access(struct g_provider *pp, int dr, int dw, int de)
{
	struct g_consumer *cp1, *cp2, *tmp;
	struct g_ventoy_disk *disk;
	struct g_geom *gp;
	int error;

    if (dw > 0) /* readonly */
		return (EPERM);

	g_topology_assert();
	gp = pp->geom;

#if 0
	/* On first open, grab an extra "exclusive" bit */
	if (pp->acr == 0 && pp->acw == 0 && pp->ace == 0)
		de++;
	/* ... and let go of it on last close */
	if ((pp->acr + dr) == 0 && (pp->acw + dw) == 0 && (pp->ace + de) == 0)
		de--;
#else
    G_DEBUG("g_ventoy_access fake de (%d)-->(0)\n", de);
    de = 0;
#endif

	LIST_FOREACH_SAFE(cp1, &gp->consumer, consumer, tmp) {
		error = g_access(cp1, dr, dw, de);
		if (error != 0)
			goto fail;
		disk = cp1->private;
		if (cp1->acr == 0 && cp1->acw == 0 && cp1->ace == 0 &&
		    disk->d_removed) {
			g_ventoy_remove_disk(disk); /* May destroy geom. */
		}
	}
	return (0);

fail:
	LIST_FOREACH(cp2, &gp->consumer, consumer) {
		if (cp1 == cp2)
			break;
		g_access(cp2, -dr, -dw, -de);
	}
	return (error);
}
```

After that rebuild the driver and test.

**Reply by probonopd** [reply]
Thank you very much. Is there any chance you could upload the compiled kernel module for FreeSBSD 13 please?

**Reply by ventoy** [reply]
Can the code work OK in your helloSystem?

**Reply by probonopd** [reply]
I can't say because I was not yet able to compile the module. Could you attach the compiled module here please? Then I could test it more easily.

**Reply by ventoy** [reply]
Try this ko
[geom_ventoy.zip](https://github.com/ventoy/Ventoy/files/7638059/geom_ventoy.zip)

**Reply by probonopd** [reply]
Thank you. Unfortunately it is not working properly for me: I made an ISO that includes this version. It boots but at the point when `/usr/local/sbin/automount` tries to mount devices, then the system becomes unresponsive and/or Xorg is restarted in an endless loop. This happens when booting the ISO from an exfat Ventoy partition on an Intel-based EFI system. Here is the ISO in case it is useful for testing/debugging: https://github.com/helloSystem/ISO/releases/tag/rw-ventoy

To see kernel messages, press the Backspace key while it says "Loading kernel...". Then, at the `OK` prompt, enter `unset boot_mute` and `boot`.

**Comment by ventoy** [comment]
`volume was not unmounted cleanly`  doesn't matter. That just means that the exfat partition was not normally unmounted last time.
